plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2536,9 +2635,232 @@ void producer_plugin::log_failed_transaction(const transaction_id_type& trx_id,
             ("entire_trx", packed_trx_ptr ? my->chain_plug->get_log_trx(packed_trx_ptr->get_transaction()) : fc::variant{trx_id}));
 }
 
-const std::set<account_name>& producer_plugin::producer_accounts() const {
-   return my->_producers;
+// Called from app thread
+void producer_plugin_impl::switch_to_write_window() {
+   EOS_ASSERT(app().executor().is_read_window(), producer_exception, "expected to be in read window");
+   EOS_ASSERT(_ro_num_active_trx_exec_tasks.load() == 0, producer_exception, "no read-only tasks should be running before switching to write window");
+
+   _ro_read_window_timer.cancel();
+   _ro_write_window_timer.cancel();
+
+   // this methond can be called from multiple places. it is possible
+   // we are already in write window.
+   if ( app().executor().is_write_window() )
+      return;
+   start_write_window();
+}
+
+// called from app thread.
+void producer_plugin_impl::start_write_window() {
+   chain::controller& chain = chain_plug->chain();
+
+   app().executor().set_to_write_window();
+   chain.unset_db_read_only_mode();
+
+   auto expire_time = boost::posix_time::microseconds(_ro_write_window_time_us.count());
+   _ro_write_window_timer.expires_from_now( expire_time );
+   _ro_write_window_timer.async_wait( app().executor().wrap(
+      priority::high,
+      exec_queue::read_only_trx_safe, // placed in read_only_trx_safe queue so it is ensured to be executed in either window
+      [weak_this = weak_from_this()]( const boost::system::error_code& ec ) {
+         auto self = weak_this.lock();
+         if( self && ec != boost::asio::error::operation_aborted ) {
+            self->switch_to_read_window();
+         }
+      }));
+}
+
+// called from app thread
+void producer_plugin_impl::switch_to_read_window() {
```

Comment:
Nothing prevents the `_timer` from firing during the read window. `_timer` will cause an `abort_block()/start_block()`

Reply:
For `_timer.async_wait( app().executor().wrap( priority::high, exec_queue::general,`, can we change `wrap` to `post` so the timer task will get executed only after the read window ends?

Reply:
Actually, `wrap` should be no different than `post` here. So this timer will fire but the lambda will not execute until you back in the write window. This can delay the production of a block by read window time. If we disable read-only completely on BP node then no issue. If we keep read-only on BP nodes then we should do something to interrupt the read-only window when this timer fires.